File: src/bot.ts
import { Client, GatewayIntentBits } from "discord.js";
import type { Message, TextChannel, ChatInputCommandInteraction } from "discord.js";
import { splitMessage } from "./discord/message-splitter.js";

const TYPING_INTERVAL_MS = 5000;

export type BotConfig = {
  token: string;
  onMessage: (content: string, channelId: string) => Promise<string>;
  onToneCommand?: (args: string) => string;
  onCalendarCommand?: (args: string, channelId: string) => Promise<string>;
  onCalendarInput?: (content: string, channelId: string) => Promise<{ handled: boolean; response: string }>;
  onChannelCommand?: (args: string, channelId: string) => string;
  onClearCommand?: (channelId: string) => string;
  isAlwaysOnChannel?: (channelId: string) => boolean;
  onScheduleCommand?: (
    sub: string,
    options: { expression?: string; prompt?: string; id?: string },
    channelId: string,
  ) => string;
};

export function createBot(config: BotConfig): Client {
  const client = new Client({
    intents: [
      GatewayIntentBits.Guilds,
      GatewayIntentBits.GuildMessages,
      GatewayIntentBits.MessageContent,
    ],
  });

  client.on("messageCreate", (message: Message) =>
    handleMessage(client, config, message),
  );

  client.on("interactionCreate", (interaction) => {
    if (!interaction.isChatInputCommand()) return;
    handleSlashCommand(config, interaction as ChatInputCommandInteraction).catch((err) => {
      process.stderr.write(`Error handling slash command: ${err instanceof Error ? err.message : String(err)}\n`);
    });
  });

  client.login(config.token);

  return client;
}

async function handleSlashCommand(
  config: BotConfig,
  interaction: ChatInputCommandInteraction,
): Promise<void> {
  const { commandName, channelId } = interaction;

  if (commandName === "clear") {
    const response = config.onClearCommand?.(channelId) ?? "clear コマンドは設定されていません。";
    await interaction.reply({ content: response });
    return;
  }

  if (commandName === "tone") {
    if (!config.onToneCommand) {
      await interaction.reply({ content: "tone コマンドは設定されていません。", ephemeral: true });
      return;
    }
    const sub = interaction.options.getSubcommand();
    let args = "";
    if (sub === "reset") args = "reset";
    else if (sub === "preset") args = interaction.options.getString("name", true);
    else if (sub === "custom") args = `set ${interaction.options.getString("text", true)}`;
    // sub === "show" → args = "" (default)
    await interaction.reply({ content: config.onToneCommand(args) });
    return;
  }

  if (commandName === "calendar") {
    await interaction.deferReply();
    const sub = interaction.options.getSubcommand();
    let args = sub;
    if (sub === "default" || sub === "default-global") {
      args = `${sub} ${interaction.options.getString("name", true)}`;
    }
    try {
      const response = await config.onCalendarCommand?.(args, channelId) ?? "";
      if (response.trim().length > 0) await interaction.editReply({ content: response });
      else await interaction.deleteReply();
    } catch {
      await interaction.editReply({ content: "カレンダーコマンドの処理中にエラーが発生しました。" });
    }
    return;
  }

  if (commandName === "channel") {
    const sub = interaction.options.getSubcommand();
    const response = config.onChannelCommand?.(sub, channelId) ?? "channel コマンドは設定されていません。";
    await interaction.reply({ content: response });
    return;
  }

  if (commandName === "schedule") {
    const sub = interaction.options.getSubcommand();
    const opts: { expression?: string; prompt?: string; id?: string } = {};
    if (sub === "add") {
      opts.expression = interaction.options.getString("expression", true);
      opts.prompt = interaction.options.getString("prompt", true);
    } else if (sub === "delete") {
      opts.id = interaction.options.getString("id", true);
    }
    const response = config.onScheduleCommand?.(sub, opts, channelId) ?? "schedule コマンドは設定されていません。";
    await interaction.reply({ content: response });
    return;
  }
}

async function handleMessage(
  client: Client,
  config: BotConfig,
  message: Message,
): Promise<void> {
  if (message.author.bot) return;
  const channel = message.channel as TextChannel;
  const isMentioned = client.user != null && message.mentions.has(client.user);
  const isAlwaysOn = config.isAlwaysOnChannel?.(channel.id) ?? false;
  if (!isMentioned && !isAlwaysOn) return;
  const prompt = isMentioned
    ? message.content.replace(/<@!?\d+>/g, "").trim()
    : message.content.trim();

  if (config.onCalendarInput) {
    try {
      const calendarResult = await config.onCalendarInput(prompt, channel.id);
      if (calendarResult.handled) {
        if (calendarResult.response.trim().length > 0) {
          await channel.send(calendarResult.response);
        }
        return;
      }
    } catch (error: unknown) {
      process.stderr.write(
        `Error handling calendar input: ${error instanceof Error ? error.message : String(error)}\n`,
      );
      await channel
        .send("カレンダー処理中にエラーが発生しました。/calendar help で使い方を確認してください。")
        .catch(() => {});
      return;
    }
  }

  const typingInterval = setInterval(() => {
    channel.sendTyping().catch(() => {});
  }, TYPING_INTERVAL_MS);

  try {
    channel.sendTyping().catch(() => {});
    const response = await config.onMessage(prompt, channel.id);
    const chunks = splitMessage(response);
    for (const chunk of chunks) {
      await channel.send(chunk);
    }
  } catch (error: unknown) {
    process.stderr.write(
      `Error handling message: ${error instanceof Error ? error.message : String(error)}\n`,
    );
    await channel.send("An error occurred while processing your message.").catch(() => {});
  } finally {
    clearInterval(typingInterval);
  }
}

File: src/bot.py
import re
import asyncio
from typing import Callable, Awaitable
import discord
from src.discord.splitter import split_message

OnMentionFn = Callable[[str, str, "list[tuple[bytes, str]] | None"], Awaitable[str]]
OnMessageFn = Callable[[object], Awaitable[None]]
OnReadyFn = Callable[[], Awaitable[None]]
OnRandomMessageFn = Callable[[object], Awaitable[None]]


class DiscordBot:
    def __init__(
        self,
        on_mention: OnMentionFn,
        on_message: OnMessageFn,
        on_ready: OnReadyFn | None = None,
        on_random_message: OnRandomMessageFn | None = None,
    ) -> None:
        intents = discord.Intents.default()
        intents.message_content = True
        self._client = discord.Client(intents=intents)
        self._on_mention = on_mention
        self._on_message_reaction = on_message
        self._on_ready_cb = on_ready
        self._on_random_message = on_random_message
        self._setup_events()

    def _setup_events(self) -> None:
        @self._client.event
        async def on_ready() -> None:
            if self._on_ready_cb:
                await self._on_ready_cb()

        @self._client.event
        async def on_message(message: discord.Message) -> None:
            await self._on_message_handler(message)

    async def _on_message_handler(self, message: discord.Message) -> None:
        if message.author.bot:
            return

        async def _run_safe(coro: object, label: str) -> None:
            try:
                await coro  # type: ignore[misc]
            except Exception as e:
                print(f"[{label}] error: {e}")

        # リアクション処理（全メッセージ対象、非同期）
        asyncio.create_task(_run_safe(self._on_message_reaction(message), "reaction"))

        # メンション処理
        if self._client.user not in message.mentions:
            if self._on_random_message:
                asyncio.create_task(_run_safe(self._on_random_message(message), "whimsy"))
            return

        prompt = re.sub(r"<@!?\d+>", "", message.content).strip()
        if not prompt:
            return

        images: list[tuple[bytes, str]] = []
        for attachment in message.attachments:
            if attachment.content_type and attachment.content_type.startswith("image/"):
                images.append((await attachment.read(), attachment.content_type))

        async with message.channel.typing():
            try:
                response = await self._on_mention(prompt, str(message.channel.id), images or None)
            except Exception as e:
                response = f"エラーが発生しました: {e}"

        for part in split_message(response):
            await message.channel.send(part)

    def run(self, token: str) -> None:
        self._client.run(token)


def create_bot(
    on_mention: OnMentionFn,
    on_message: OnMessageFn,
    on_ready: OnReadyFn | None = None,
    on_random_message: OnRandomMessageFn | None = None,
) -> DiscordBot:
    return DiscordBot(
        on_mention=on_mention,
        on_message=on_message,
        on_ready=on_ready,
        on_random_message=on_random_message,
    )
